### PR TITLE
[Bugfix] Fix Ray Metrics API usage

### PR DESCRIPTION
**Pull request by Yard1**
Fixes `RayMetrics` not working correctly due to wrong API usage.

Fixes https://github.com/vllm-project/vllm/pull/5925#issuecomment-2223941096 cc @kousun12
---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/engine/metrics.py [hunk 1]
```diff
@@ -168,8 +246,9 @@ def _unregister_vllm_metrics(self) -> None:
         # No-op on purpose
         pass
 
-
-# end-metrics-definitions
+    def _create_info_cache_config(self) -> None:
```
Comment (kousun12) [inline vllm/engine/metrics.py:249]: thanks for fixing this as well!

**Review by kousun12 — APPROVED** [review]
looks great. thank you!

File: vllm/engine/metrics.py [hunk 2]
```diff
@@ -97,67 +97,145 @@ def __init__(self, labelnames: List[str], max_model_len: int):
...
+class _RayCounterWrapper:
+    """Wraps around ray.util.metrics.Counter to provide same API as
+    prometheus_client.Counter"""
+
+    def __init__(self,
+                 name: str,
+                 documentation: str = "",
+                 labelnames: Optional[List[str]] = None):
+        labelnames_tuple = tuple(labelnames) if labelnames else None
+        self._counter = ray_metrics.Counter(name=name,
```
Comment (bchandarr) [inline vllm/engine/metrics.py:194]: the Counter class (https://docs.ray.io/en/latest/_modules/ray/util/metrics.html) has the check value <= 0 which also fails as the value starts with 0.

**Comment by bchandarr** [comment]
With the new stat_loggers param in LLMEngine, I get serialise error (in ray remote/ray actor) with passing the dict of RayPrometheusStatLogger object. Can this be updated to specify "ray" or "prometheus" as input and create PrometheusStatLogger or RayPrometheusStatLogger?

File: vllm/engine/metrics.py (hunk 2 above)
Comment (Yard1) [reply vllm/engine/metrics.py:194]: Line 203 in this file should take care of this

**Comment by Yard1** [comment]
@bchandarr do you have a repro?

**Review by rkooo567 — APPROVED** [review]

**Comment by bchandarr** [comment]
> @bchandarr do you have a repro?

I use custom AsyncLLMEngine with from_engine_args function and tried like the code below, not sure this is the right way to use it.
```
  def from_engine_args(
      cls,
      engine_args: AsyncEngineArgs,
      start_engine_loop: bool = True,
      usage_context: UsageContext = UsageContext.ENGINE_CONTEXT,
  ) -> "AsyncLLMEngine":

     # rest of the original code here

      # create loggers
      stat_loggers = {
          "raymetrics":
          RayPrometheusStatLogger(
              local_interval=_LOCAL_LOGGING_INTERVAL_SEC,
              labels=dict(model_name=model_config.served_model_name),
              max_model_len=model_config.max_model_len),
      }

      # Create the async LLM engine.
      engine = cls(
          distributed_executor_backend == "ray",
          # parallel_config.worker_use_ray,
          engine_args.engine_use_ray,
          **engine_config.to_dict(),
          executor_class=executor_class,
          log_requests=not engine_args.disable_log_requests,
          log_stats=not engine_args.disable_log_stats,
          max_log_len=engine_args.max_log_len,
          start_engine_loop=start_engine_loop,
          usage_context=usage_context,
          stat_loggers=stat_loggers,
      )
      return engine
```
the error call stack.
```
  File "/home/ray/projects/llm_serve/app/backends/vllm/vllm_engine.py", line 131, in from_engine_args
    engine = cls(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/vllm/engine/async_llm_engine.py", line 360, in __init__
    self.engine = self._init_engine(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/vllm/engine/async_llm_engine.py", line 507, in _init_engine
    return engine_class(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/actor.py", line 715, in remote
    return self._remote(args=args, kwargs=kwargs, **self._default_options)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/actor.py", line 1161, in _remote
    actor_id = worker.core_worker.create_actor(
TypeError: Could not serialize the argument {'raymetrics': <llm_serve.app.backends.vllm.metrics.RayPrometheusStatLogger object at 0x7f008e183d90>} for a task or actor llm_serve.app.backends.vllm.vllm_engine.CustomLLMEngine.__init__:
================================================================================
Checking Serializability of {'raymetrics': <llm_serve.app.backends.vllm.metrics.RayPrometheusStatLogger object at 0x7f008e183d90>}
================================================================================
[31m!!! FAIL[39m serialization: cannot pickle '_thread.lock' object
WARNING: Did not find non-serializable object in {'raymetrics': <llm_serve.app.backends.vllm.metrics.RayPrometheusStatLogger object at 0x7f008e183d90>}. This may be an oversight.
================================================================================
Variable: 

	[1mFailTuple({'raymetrics': <llm_serve.app.backends.vllm.metrics.RayPrometheusStatLogger object at 0x7f008e183d90>} [obj={'raymetrics': <llm_serve.app.backends.vllm.metrics.RayPrometheusStatLogger object at 0x7f008e183d90>}, parent=None])[0m

was found to be non-serializable. There may be multiple other undetected variables that were non-serializable. 
Consider either removing the instantiation/imports of these variables or moving the instantiation into the scope of the function/class. 
================================================================================
Check https://docs.ray.io/en/master/ray-core/objects/serialization.html#troubleshooting for more information.
If you have any suggestions on how to improve this error message, please reach out to the Ray developers on github.com/ray-project/ray/issues/
================================================================================

```

**Comment by bchandarr** [comment]
@Yard1 Also I tried running the test script in v0.5.1 and get error from prometheus_client/registry.py

```
import ray
from vllm import EngineArgs, LLMEngine
from vllm.engine.arg_utils import AsyncEngineArgs
from vllm.engine.async_llm_engine import AsyncLLMEngine
from vllm.engine.metrics import RayPrometheusStatLogger
from vllm.sampling_params import SamplingParams

def test_engine_log_metrics_ray(
    example_prompts = ["prompt"],
    model: str = “some-model",
    dtype: str = "half",
    max_tokens: int = 16,
) -> None:
    # This test is quite weak - it only checks that we can use
    # RayPrometheusStatLogger without exceptions.
    # Checking whether the metrics are actually emitted is unfortunately
    # non-trivial.

    # We have to run in a Ray task for Ray metrics to be emitted correctly
    @ray.remote(num_gpus=1)
    def _inner():

        class _RayPrometheusStatLogger(RayPrometheusStatLogger):

            def __init__(self, *args, **kwargs):
                self._i = 0
                super().__init__(*args, **kwargs)

            def log(self, *args, **kwargs):
                self._i += 1
                return super().log(*args, **kwargs)

        engine_args = EngineArgs(
            model=model,
            dtype=dtype,
            disable_log_stats=False,
        )
        engine = LLMEngine.from_engine_args(engine_args)
        logger = _RayPrometheusStatLogger(
            local_interval=0.5,
            labels=dict(model_name=engine.model_config.served_model_name),
            max_model_len=engine.model_config.max_model_len)
        engine.add_logger("ray", logger)
        for i, prompt in enumerate(example_prompts):
            engine.add_request(
                f"request-id-{i}",
                prompt,
                SamplingParams(max_tokens=max_tokens),
            )
        while engine.has_unfinished_requests():
            engine.step()
        assert logger._i > 0, ".log must be called at least once"

    ray.get(_inner.remote())

test_engine_log_metrics_ray()
```



```
(venv) ray@raycluster:~/$ python test.py
(_inner pid=11587) INFO llm_engine.py:169] Initializing an LLM engine (v0.5.1) with config: model=‘some-model’, speculative_config=None, tokenizer=‘some-modelchat', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.float16, max_seq_len=4096, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None), seed=0, served_model_name=some-model, use_v2_block_manager=False, enable_prefix_caching=False)
(_inner pid=11587) INFO  model_runner.py:255] Loading model weights took 12.5523 GB
(_inner pid=11587) INFO  gpu_executor.py:84] # GPU blocks: 7447, # CPU blocks: 512
(_inner pid=11587) INFO  model_runner.py:924] Capturing the model for CUDA graphs. This may lead to unexpected consequences if the model is not static. To run the model in eager mode, set 'enforce_eager=True' or use '--enforce-eager' in the CLI.
(_inner pid=11587) INFO  model_runner.py:928] CUDA graphs can take additional 1~3 GiB memory per GPU. If you are running out of memory, consider decreasing `gpu_memory_utilization` or enforcing eager mode. You can also reduce the `max_num_seqs` as needed to decrease memory usage.
(_inner pid=11587) INFO  model_runner.py:1117] Graph capturing finished in 5 secs.
Traceback (most recent call last):
  File "/home/ray/test.py", line 60, in <module>
    test_engine_log_metrics_ray()
  File "/home/ray/test.py", line 58, in test_engine_log_metrics_ray
    ray.get(_inner.remote())
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/worker.py", line 2623, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/worker.py", line 861, in get_objects
    raise value.as_instanceof_cause()
ray.exceptions.RayTaskError(ValueError): ray::_inner() (pid=11587, ip=192.128.18.68)
  File "/home/ray/projects/llm_serve/python/test.py", line 43, in _inner
    logger = _RayPrometheusStatLogger(
  File "/home/ray/projects/llm_serve/python/test.py", line 31, in __init__
    super().__init__(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/vllm/engine/metrics.py", line 341, in __init__
    self.metrics = self._metrics_cls(labelnames=list(labels.keys()),
  File "/home/ray/anaconda3/lib/python3.10/site-packages/vllm/engine/metrics.py", line 165, in __init__
    super().__init__(labelnames, max_model_len)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/vllm/engine/metrics.py", line 40, in __init__
    self.info_cache_config = prometheus_client.Info(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/prometheus_client/metrics.py", line 155, in __init__
    registry.register(self)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/prometheus_client/registry.py", line 43, in register
    raise ValueError(
ValueError: Duplicated timeseries in CollectorRegistry: {'vllm:cache_config', 'vllm:cache_config_info'}
```

**Comment by Yard1** [comment]
thanks will take a look

**Comment by Yard1** [comment]
@bchandarr I cannot reproduce either issue. The logger is serializable for me, and I am not getting the error you got in the second issue. Not sure what's going on here.

**Pull request merged**